OrbitCaptureClient/CaptureEventProcessor.cpp

```diff
@@ -340,3 +423,294 @@ void CaptureEventProcessor::SendTracepointInfoToListenerIfNecessary(
+  if (gpu_job_matching_pre_submission_it == submission_time_to_gpu_job.end()) {
+    return nullptr;
+  }
+
+  // Find the first Gpu job that has a timestamp greater to the "post submission" timestamp
+  // (which would be the next job) and decrease the iterator by one.
+  auto gpu_job_matching_post_submission_it =
+      submission_time_to_gpu_job.upper_bound(post_submission_cpu_timestamp);
+  if (gpu_job_matching_post_submission_it == submission_time_to_gpu_job.begin()) {
+    return nullptr;
+  }
+  --gpu_job_matching_post_submission_it;
+
+  if (&gpu_job_matching_pre_submission_it->second != &gpu_job_matching_post_submission_it->second) {
+    return nullptr;
+  }
+
+  return &gpu_job_matching_pre_submission_it->second;
+}
+
+void CaptureEventProcessor::ProcessGpuQueueSubmissionWithMatchingGpuJob(
+    const GpuQueueSubmission& gpu_queue_submission, const GpuJob& matching_gpu_job) {
+  std::string timeline;
+  if (matching_gpu_job.timeline_or_key_case() == GpuJob::kTimelineKey) {
+    CHECK(string_intern_pool_.contains(matching_gpu_job.timeline_key()));
+    timeline = string_intern_pool_.at(matching_gpu_job.timeline_key());
+  } else {
+    timeline = matching_gpu_job.timeline();
+  }
+  uint64_t timeline_hash = GetStringHashAndSendToListenerIfNecessary(timeline);
+
+  std::optional<GpuCommandBuffer> first_command_buffer =
+      ExtractFirstCommandBuffer(gpu_queue_submission);
+
+  ProcessGpuCommandBuffers(gpu_queue_submission, matching_gpu_job, first_command_buffer,
+                           timeline_hash);
+
+  ProcessGpuDebugMarkers(gpu_queue_submission, matching_gpu_job, first_command_buffer, timeline);
+}
+
+bool CaptureEventProcessor::HasUnprocessedBeginMarkers(int32_t thread_id,
+                                                       uint64_t post_submission_timestamp) const {
+  if (!tid_to_post_submission_time_to_num_begin_markers_.contains(thread_id)) {
+    return false;
+  }
+  if (!tid_to_post_submission_time_to_num_begin_markers_.at(thread_id).contains(
+          post_submission_timestamp)) {
+    return false;
+  }
+  CHECK(tid_to_post_submission_time_to_num_begin_markers_.at(thread_id).at(
+            post_submission_timestamp) > 0);
+  return true;
+}
+
+void CaptureEventProcessor::DecrementUnprocessedBeginMarkers(int32_t thread_id,
+                                                             uint64_t submission_timestamp,
+                                                             uint64_t post_submission_timestamp) {
+  CHECK(tid_to_post_submission_time_to_num_begin_markers_.contains(thread_id));
+  auto& post_submission_time_to_num_begin_markers =
+      tid_to_post_submission_time_to_num_begin_markers_.at(thread_id);
+  CHECK(post_submission_time_to_num_begin_markers.contains(post_submission_timestamp));
+  uint64_t new_num = post_submission_time_to_num_begin_markers.at(post_submission_timestamp) - 1;
+  post_submission_time_to_num_begin_markers[post_submission_timestamp] = new_num;
```

Comment:
Related to the outdated comment about using `.at`, why can't you just write `--post_submission_time_to_num_begin_markers.at(post_submission_timestamp)`?

`.at` also returns a reference (or what am I missing?).

Maybe it's clearer now without the pre-increment, but you should still be able to write
`post_submission_time_to_num_begin_markers.at(post_submission_timestamp) = new_num;`.
It doesn't make a big difference but to me using `at` makes it clearer that we know and expect that the key-value pair was already in the map.

Reply:
Right. I somehow forgot about `.at` also returning a reference. Updated the code.